src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelSparkUtils.java

```diff
@@ -0,0 +1,151 @@
+        for (int begIndex = 0; begIndex < length; begIndex += blockSize) {
+            blocks.add(new LinearSpaceBlock(begIndex, FastMath.min(begIndex + blockSize, length)));
+        }
+        /* the last block might be smaller than minBlockSize; we merge them */
+        while (blocks.size() > numBlocks && blocks.size() > 1) {
+            final int newBegIndex = blocks.get(blocks.size() - 2).getBegIndex();
+            final int newEndIndex = blocks.get(blocks.size() - 1).getEndIndex();
+            blocks.remove(blocks.size() - 1);
+            blocks.remove(blocks.size() - 1);
+            blocks.add(new LinearSpaceBlock(newBegIndex, newEndIndex));
+        }
+        return blocks;
+    }
+
+    /**
+     * Partitions an {@link INDArray} along the 0th axis according to a given partitioning and returns
+     * a list
+     *
+     * @param blocks partitioning
+     * @param arr an array
+     * @return list of partitioned arrays
+     */
+    public static List<Tuple2<LinearSpaceBlock, INDArray>> partitionINDArrayToAList(@Nonnull final List<LinearSpaceBlock> blocks,
+                                                                                    @Nonnull final INDArray arr) {
+        return blocks.stream().map(block ->
+                new Tuple2<>(block, arr.get(NDArrayIndex.interval(block.getBegIndex(), block.getEndIndex()))))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Partitions an {@link INDArray} along the 0th axis according to a given partitioning and returns
+     * a map
+     *
+     * @param blocks partitioning
+     * @param arr an array
+     * @return map of partitions to partitioned arrays
+     */
+    public static Map<LinearSpaceBlock, INDArray> partitionINDArrayToAMap(@Nonnull final List<LinearSpaceBlock> blocks,
+                                                                          @Nonnull final INDArray arr) {
+        return blocks.stream().map(block ->
+                new Tuple2<>(block, arr.get(NDArrayIndex.interval(block.getBegIndex(), block.getEndIndex()))))
+                .collect(Collectors.toMap(p -> p._1, p -> p._2));
+    }
+
+    /**
+     * Assembles INDArray blocks in a {@code JavaPairRDD<LinearSpaceBlock, INDArray>} by collecting
+     * them to a local list, sorting them based on their keys ({@link LinearSpaceBlock}), and
+     * concatenating them along a given axis.
+     *
+     * @param blocksPairRDD an instance of {@code JavaPairRDD<LinearSpaceBlock, INDArray>}
+     * @param axis axis to concat along
+     * @return an instance of {@link INDArray}
+     */
+    public static INDArray assembleINDArrayBlocksFromRDD(@Nonnull JavaPairRDD<LinearSpaceBlock, INDArray> blocksPairRDD,
+                                                         final int axis) {
+        final List<INDArray> sortedBlocks = blocksPairRDD.collect().stream()
+                /* sort according to begin index of each target-space partition */
```

Comment:
Sure, but Scala-style functional code can get so mysterious so soon; see `CoverageModelEMWorkspace.updateCopyRatioPosteriorExpectationsSpark`. I'd rather keep the natural language comments to help future developers, and myself.